### ISSUE TITLE
Image scan errors on 0.24.3

### ISSUE THREAD

**Issue by ross211** [aquasecurity/trivy#1848]
## Description

Today I got a scan error on a freshly built local Docker image. This had previously worked fine, it appears related to 0.24.3 as I cannot reproduce the error if I rollback to 0.24.2.

I was able to reproduce on a few systems with the following:
```
image=index.docker.io/nginx

docker build --pull -t "$image" - <<< "FROM nginx:alpine-perl"
docker run --rm -it -v /var/run/docker.sock:/var/run/docker.sock aquasec/trivy:0.24.3 image "$image"
# this works

# now rebuild the image again with a different base image
docker build --pull -t "$image" - <<< "FROM nginx:alpine"
docker run --rm -it -v /var/run/docker.sock:/var/run/docker.sock aquasec/trivy:0.24.3 image "$image"
# this errors

# or modify the original image
docker build --pull -t "$image" - <<EOF
FROM nginx:alpine-perl
RUN apk --no-cache upgrade
EOF
docker run --rm -it -v /var/run/docker.sock:/var/run/docker.sock aquasec/trivy:0.24.3 image "$image"
# this errors too

# retry with 0.24.2
docker run --rm -it -v /var/run/docker.sock:/var/run/docker.sock aquasec/trivy:0.24.2 image "$image"
# this works
```

## What did you expect to happen?

```
2022-03-17T08:45:49.630Z	INFO	Need to update DB
2022-03-17T08:45:49.630Z	INFO	Downloading DB...
29.98 MiB / 29.98 MiB [-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% 33.23 MiB p/s 1.1s
2022-03-17T08:45:53.554Z	INFO	Detected OS: alpine
2022-03-17T08:45:53.554Z	INFO	Detecting Alpine vulnerabilities...
2022-03-17T08:45:53.556Z	INFO	Number of language-specific files: 0

index.docker.io/nginx (alpine 3.15.0)
=====================================
Total: 4 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 4, CRITICAL: 0)

+--------------+------------------+----------+-------------------+---------------+---------------------------------------+
|   LIBRARY    | VULNERABILITY ID | SEVERITY | INSTALLED VERSION | FIXED VERSION |                 TITLE                 |
+--------------+------------------+----------+-------------------+---------------+---------------------------------------+
| libcrypto1.1 | CVE-2022-0778    | HIGH     | 1.1.1l-r7         | 1.1.1n-r0     | openssl: Infinite loop in             |
|              |                  |          |                   |               | BN_mod_sqrt() reachable               |
|              |                  |          |                   |               | when parsing certificates             |
|              |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2022-0778  |
+--------------+                  +          +                   +               +                                       +
| libssl1.1    |                  |          |                   |               |                                       |
|              |                  |          |                   |               |                                       |
|              |                  |          |                   |               |                                       |
|              |                  |          |                   |               |                                       |
+--------------+------------------+          +-------------------+---------------+---------------------------------------+
| libxml2      | CVE-2022-23308   |          | 2.9.12-r2         | 2.9.13-r0     | libxml2: Use-after-free               |
|              |                  |          |                   |               | of ID and IDREF attributes            |
|              |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2022-23308 |
+--------------+------------------+          +-------------------+---------------+---------------------------------------+
| libxslt      | CVE-2021-30560   |          | 1.1.34-r1         | 1.1.35-r0     | Use after free in Blink               |
|              |                  |          |                   |               | XSLT in Google Chrome                 |
|              |                  |          |                   |               | prior to 91.0.4472.164...             |
|              |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2021-30560 |
+--------------+------------------+----------+-------------------+---------------+---------------------------------------+
```

## What happened instead?

```
2022-03-17T08:45:22.233Z	INFO	Need to update DB
2022-03-17T08:45:22.233Z	INFO	Downloading DB...
29.98 MiB / 29.98 MiB [-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% 32.53 MiB p/s 1.1s
2022-03-17T08:45:26.775Z	FATAL	scan error: image scan failed: failed analysis: analyze error: failed to analyze layer: sha256:67bae81de3dc5ccf5101dd4973e8e4b10ee3d5681feb9c30119e6acee1e12dd5 : unable to get uncompressed layer sha256:67bae81de3dc5ccf5101dd4973e8e4b10ee3d5681feb9c30119e6acee1e12dd5: failed to get the layer (sha256:67bae81de3dc5ccf5101dd4973e8e4b10ee3d5681feb9c30119e6acee1e12dd5): unable to populate: unable to open: failed to initialize the struct from the temporary file: tarball must contain only a single image to be used with tarball.Image
```

## Output of run with `-debug`:

```
# 0.24.3
ubuntu@ip-172-23-2-72:~$ docker run --rm -it -v /var/run/docker.sock:/var/run/docker.sock aquasec/trivy:0.24.3 --debug image "$image"
2022-03-17T08:49:29.186Z	DEBUG	Severities: UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL
2022-03-17T08:49:29.192Z	DEBUG	cache dir:  /root/.cache/trivy
2022-03-17T08:49:29.192Z	DEBUG	There is no valid metadata file: unable to open a file: open /root/.cache/trivy/db/metadata.json: no such file or directory
2022-03-17T08:49:29.192Z	INFO	Need to update DB
2022-03-17T08:49:29.192Z	INFO	Downloading DB...
2022-03-17T08:49:29.192Z	DEBUG	no metadata file
29.98 MiB / 29.98 MiB [-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% 33.90 MiB p/s 1.1s
2022-03-17T08:49:31.309Z	DEBUG	Updating database metadata...
2022-03-17T08:49:31.310Z	DEBUG	DB Schema: 2, UpdatedAt: 2022-03-17 06:06:51.817102324 +0000 UTC, NextUpdate: 2022-03-17 12:06:51.817101423 +0000 UTC, DownloadedAt: 2022-03-17 08:49:31.309780732 +0000 UTC
2022-03-17T08:49:31.310Z	DEBUG	Vulnerability type:  [os library]
2022-03-17T08:49:31.313Z	DEBUG	Image ID: sha256:bef258acf10dc257d641c47c3a600c92f87be4b4ce4a5e4752b3eade7533dcd9
2022-03-17T08:49:31.313Z	DEBUG	Diff IDs: [sha256:8d3ac3489996423f53d6087c81180006263b79f206d3fdec9e66f0e27ceb8759 sha256:67bae81de3dc5ccf5101dd4973e8e4b10ee3d5681feb9c30119e6acee1e12dd5 sha256:89f4d03665cee1ca943d0b55c086ea7577f4658c2cabe69eaf640f1dc47129f9 sha256:318191938fd7801d80ad617482e4312f0ced71e6c1d4bcb02e12d90a9d65eb11 sha256:a770f8eba3cb3840d7579e2c4925c9fa275813e8719fe5435905b887a9ee7674 sha256:6fda88393b8b02872c87c6626807e7bfac1f49e63f47a5ddba0cd11fae99b575]
2022-03-17T08:49:31.314Z	DEBUG	Missing image ID: sha256:bef258acf10dc257d641c47c3a600c92f87be4b4ce4a5e4752b3eade7533dcd9
2022-03-17T08:49:31.314Z	DEBUG	Missing diff ID: sha256:6fda88393b8b02872c87c6626807e7bfac1f49e63f47a5ddba0cd11fae99b575
2022-03-17T08:49:31.314Z	DEBUG	Missing diff ID: sha256:8d3ac3489996423f53d6087c81180006263b79f206d3fdec9e66f0e27ceb8759
2022-03-17T08:49:31.314Z	DEBUG	Missing diff ID: sha256:67bae81de3dc5ccf5101dd4973e8e4b10ee3d5681feb9c30119e6acee1e12dd5
2022-03-17T08:49:31.314Z	DEBUG	Missing diff ID: sha256:89f4d03665cee1ca943d0b55c086ea7577f4658c2cabe69eaf640f1dc47129f9
2022-03-17T08:49:31.314Z	DEBUG	Missing diff ID: sha256:318191938fd7801d80ad617482e4312f0ced71e6c1d4bcb02e12d90a9d65eb11
2022-03-17T08:49:31.314Z	DEBUG	Missing diff ID: sha256:a770f8eba3cb3840d7579e2c4925c9fa275813e8719fe5435905b887a9ee7674
2022-03-17T08:49:33.946Z	FATAL	scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.runWithTimeout
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:70
  - image scan failed:
    github.com/aquasecurity/trivy/pkg/commands/artifact.scan
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:219
  - failed analysis:
    github.com/aquasecurity/trivy/pkg/scanner.Scanner.ScanArtifact
        /home/runner/work/trivy/trivy/pkg/scanner/scan.go:96
  - analyze error:
    github.com/aquasecurity/fanal/artifact/image.Artifact.Inspect
        /home/runner/go/pkg/mod/github.com/aquasecurity/fanal@v0.0.0-20220309115321-0183e4870f89/artifact/image/image.go:105
  - failed to analyze layer: sha256:6fda88393b8b02872c87c6626807e7bfac1f49e63f47a5ddba0cd11fae99b575 :
    github.com/aquasecurity/fanal/artifact/image.Artifact.inspect.func1
        /home/runner/go/pkg/mod/github.com/aquasecurity/fanal@v0.0.0-20220309115321-0183e4870f89/artifact/image/image.go:154
  - unable to get uncompressed layer sha256:6fda88393b8b02872c87c6626807e7bfac1f49e63f47a5ddba0cd11fae99b575:
    github.com/aquasecurity/fanal/artifact/image.Artifact.inspectLayer
        /home/runner/go/pkg/mod/github.com/aquasecurity/fanal@v0.0.0-20220309115321-0183e4870f89/artifact/image/image.go:194
  - failed to get the layer (sha256:6fda88393b8b02872c87c6626807e7bfac1f49e63f47a5ddba0cd11fae99b575):
    github.com/aquasecurity/fanal/artifact/image.Artifact.uncompressedLayer
        /home/runner/go/pkg/mod/github.com/aquasecurity/fanal@v0.0.0-20220309115321-0183e4870f89/artifact/image/image.go:251
  - unable to populate:
    github.com/aquasecurity/fanal/image/daemon.(*image).LayerByDiffID
        /home/runner/go/pkg/mod/github.com/aquasecurity/fanal@v0.0.0-20220309115321-0183e4870f89/image/daemon/image.go:125
  - unable to open:
    github.com/aquasecurity/fanal/image/daemon.(*image).populateImage
        /home/runner/go/pkg/mod/github.com/aquasecurity/fanal@v0.0.0-20220309115321-0183e4870f89/image/daemon/image.go:78
  - failed to initialize the struct from the temporary file:
    github.com/aquasecurity/fanal/image/daemon.imageOpener.func1
        /home/runner/go/pkg/mod/github.com/aquasecurity/fanal@v0.0.0-20220309115321-0183e4870f89/image/daemon/image.go:46
  - tarball must contain only a single image to be used with tarball.Image

# 0.24.2
ubuntu@ip-172-23-2-72:~$ docker run --rm -it -v /var/run/docker.sock:/var/run/docker.sock aquasec/trivy:0.24.2 --debug image "$image"
2022-03-17T08:49:45.146Z	DEBUG	Severities: UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL
2022-03-17T08:49:45.152Z	DEBUG	cache dir:  /root/.cache/trivy
2022-03-17T08:49:45.152Z	DEBUG	There is no valid metadata file: unable to open a file: open /root/.cache/trivy/db/metadata.json: no such file or directory
2022-03-17T08:49:45.152Z	INFO	Need to update DB
2022-03-17T08:49:45.152Z	INFO	Downloading DB...
2022-03-17T08:49:45.152Z	DEBUG	no metadata file
29.98 MiB / 29.98 MiB [-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% 34.08 MiB p/s 1.1s
2022-03-17T08:49:47.304Z	DEBUG	Updating database metadata...
2022-03-17T08:49:47.304Z	DEBUG	DB Schema: 2, UpdatedAt: 2022-03-17 06:06:51.817102324 +0000 UTC, NextUpdate: 2022-03-17 12:06:51.817101423 +0000 UTC, DownloadedAt: 2022-03-17 08:49:47.304518922 +0000 UTC
2022-03-17T08:49:47.304Z	DEBUG	Vulnerability type:  [os library]
2022-03-17T08:49:47.306Z	DEBUG	Image ID: sha256:bef258acf10dc257d641c47c3a600c92f87be4b4ce4a5e4752b3eade7533dcd9
2022-03-17T08:49:47.306Z	DEBUG	Diff IDs: [sha256:8d3ac3489996423f53d6087c81180006263b79f206d3fdec9e66f0e27ceb8759 sha256:67bae81de3dc5ccf5101dd4973e8e4b10ee3d5681feb9c30119e6acee1e12dd5 sha256:89f4d03665cee1ca943d0b55c086ea7577f4658c2cabe69eaf640f1dc47129f9 sha256:318191938fd7801d80ad617482e4312f0ced71e6c1d4bcb02e12d90a9d65eb11 sha256:a770f8eba3cb3840d7579e2c4925c9fa275813e8719fe5435905b887a9ee7674 sha256:6fda88393b8b02872c87c6626807e7bfac1f49e63f47a5ddba0cd11fae99b575]
2022-03-17T08:49:47.307Z	DEBUG	Missing image ID: sha256:bef258acf10dc257d641c47c3a600c92f87be4b4ce4a5e4752b3eade7533dcd9
2022-03-17T08:49:47.307Z	DEBUG	Missing diff ID: sha256:6fda88393b8b02872c87c6626807e7bfac1f49e63f47a5ddba0cd11fae99b575
2022-03-17T08:49:47.307Z	DEBUG	Missing diff ID: sha256:318191938fd7801d80ad617482e4312f0ced71e6c1d4bcb02e12d90a9d65eb11
2022-03-17T08:49:47.307Z	DEBUG	Missing diff ID: sha256:89f4d03665cee1ca943d0b55c086ea7577f4658c2cabe69eaf640f1dc47129f9
2022-03-17T08:49:47.307Z	DEBUG	Missing diff ID: sha256:a770f8eba3cb3840d7579e2c4925c9fa275813e8719fe5435905b887a9ee7674
2022-03-17T08:49:47.307Z	DEBUG	Missing diff ID: sha256:8d3ac3489996423f53d6087c81180006263b79f206d3fdec9e66f0e27ceb8759
2022-03-17T08:49:47.307Z	DEBUG	Missing diff ID: sha256:67bae81de3dc5ccf5101dd4973e8e4b10ee3d5681feb9c30119e6acee1e12dd5
2022-03-17T08:49:49.061Z	DEBUG	Missing image cache: sha256:fb28af7555e7690dab6cab975fc8e3e33a65c5d1828de8df03680f4d9c312e6e
2022-03-17T08:49:49.062Z	INFO	Detected OS: alpine
2022-03-17T08:49:49.062Z	INFO	Detecting Alpine vulnerabilities...
2022-03-17T08:49:49.062Z	DEBUG	alpine: os version: 3.15
2022-03-17T08:49:49.062Z	DEBUG	alpine: the number of packages: 42
2022-03-17T08:49:49.065Z	INFO	Number of language-specific files: 0

index.docker.io/nginx (alpine 3.15.0)
=====================================
Total: 4 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 4, CRITICAL: 0)

+--------------+------------------+----------+-------------------+---------------+---------------------------------------+
|   LIBRARY    | VULNERABILITY ID | SEVERITY | INSTALLED VERSION | FIXED VERSION |                 TITLE                 |
+--------------+------------------+----------+-------------------+---------------+---------------------------------------+
| libcrypto1.1 | CVE-2022-0778    | HIGH     | 1.1.1l-r7         | 1.1.1n-r0     | openssl: Infinite loop in             |
|              |                  |          |                   |               | BN_mod_sqrt() reachable               |
|              |                  |          |                   |               | when parsing certificates             |
|              |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2022-0778  |
+--------------+                  +          +                   +               +                                       +
| libssl1.1    |                  |          |                   |               |                                       |
|              |                  |          |                   |               |                                       |
|              |                  |          |                   |               |                                       |
|              |                  |          |                   |               |                                       |
+--------------+------------------+          +-------------------+---------------+---------------------------------------+
| libxml2      | CVE-2022-23308   |          | 2.9.12-r2         | 2.9.13-r0     | libxml2: Use-after-free               |
|              |                  |          |                   |               | of ID and IDREF attributes            |
|              |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2022-23308 |
+--------------+------------------+          +-------------------+---------------+---------------------------------------+
| libxslt      | CVE-2021-30560   |          | 1.1.34-r1         | 1.1.35-r0     | Use after free in Blink               |
|              |                  |          |                   |               | XSLT in Google Chrome                 |
|              |                  |          |                   |               | prior to 91.0.4472.164...             |
|              |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2021-30560 |
+--------------+------------------+----------+-------------------+---------------+---------------------------------------+
```

## Output of `trivy -v`:

```
ubuntu@ip-172-23-2-72:~$ docker run --rm -it -v /var/run/docker.sock:/var/run/docker.sock aquasec/trivy:0.24.3 -v
Version: 0.24.3
ubuntu@ip-172-23-2-72:~$ docker run --rm -it -v /var/run/docker.sock:/var/run/docker.sock aquasec/trivy:0.24.2 -v
Version: 0.24.2
```

**Comment by afdesk** [comment]
@ross211 I could reproduce this issue. 
`trivy` can't select an image correctly. I need to investigate it more.
as a workaround - you can set a tag for your image:
```sh
image=index.docker.io/nginx:latest
```

**Comment by afdesk** [comment]
@ross211 thanks again! 
`trivy` v.24.4 should fix this issue

**Comment by ross211** [comment]
Thanks for the quick fix, I can confirm 0.24.4 resolved it.
